lib/iris/common/resolve.py

```diff
@@ -0,0 +1,1541 @@
+        self.lhs_cube_dim_coverage = None
+        # Analysis of aux and scalar coordinates spanning LHS cube.
+        self.lhs_cube_aux_coverage = None
+        # Analysis of dim coordinates spanning RHS cube.
+        self.rhs_cube_dim_coverage = None
+        # Analysis of aux and scalar coordinates spanning RHS cube.
+        self.rhs_cube_aux_coverage = None
+
+        # Map common metadata from RHS cube to LHS cube if LHS-rank >= RHS-rank,
+        # otherwise map common metadata from LHS cube to RHS cube.
+        if self.lhs_cube.ndim >= self.rhs_cube.ndim:
+            self.map_rhs_to_lhs = True
+        else:
+            self.map_rhs_to_lhs = False
+
+        # Mapping of the dimensions between common metadata for the cubes,
+        # where the direction of the mapping is governed by map_rhs_to_lhs.
+        self.mapping = None
+
+        # Cache containing a list of dim, aux and scalar coordinates prepared
+        # and ready for creating and attaching to the resultant cube.
+        self.prepared_category = None
+
+        # Cache containing a list of aux factories prepared and ready for
+        # creating and attaching to the resultant cube.
+        self.prepared_factories = None
+
+        # The shape of the resultant resolved cube.
+        self._broadcast_shape = None
+
+    def _metadata_coverage(self):
+        # Determine the common dim coordinate metadata coverage.
+        common_dim_metadata = [
+            item.metadata for item in self.category_common.items_dim
+        ]
+
+        self.lhs_cube_dim_coverage = self._dim_coverage(
+            self.lhs_cube,
+            self.lhs_cube_category.items_dim,
+            common_dim_metadata,
+        )
+        self.rhs_cube_dim_coverage = self._dim_coverage(
+            self.rhs_cube,
+            self.rhs_cube_category.items_dim,
+            common_dim_metadata,
+        )
+
+        # Determine the common aux and scalar coordinate metadata coverage.
+        common_aux_metadata = [
+            item.metadata for item in self.category_common.items_aux
+        ]
+        common_scalar_metadata = [
+            item.metadata for item in self.category_common.items_scalar
+        ]
+
+        self.lhs_cube_aux_coverage = self._aux_coverage(
+            self.lhs_cube,
+            self.lhs_cube_category.items_aux,
+            self.lhs_cube_category.items_scalar,
+            common_aux_metadata,
+            common_scalar_metadata,
+        )
+        self.rhs_cube_aux_coverage = self._aux_coverage(
+            self.rhs_cube,
+            self.rhs_cube_category.items_aux,
+            self.rhs_cube_category.items_scalar,
+            common_aux_metadata,
+            common_scalar_metadata,
+        )
+
+    def _metadata_mapping(self):
+        # Initialise the state.
+        self.mapping = {}
+
+        # Map RHS cube to LHS cube, or smaller to larger cube rank.
+        if self.map_rhs_to_lhs:
+            src_cube = self.rhs_cube
```

Comment:
@lbdreyer That's understandable. I totally get that...

But these concepts are pretty core to `Resolve`, and it took me a while to settle on the best way to do this. But there are two important concepts at play here, that are subtly related, but different.

Essentially, the `lhs` and `rhs` refer to the cube operands, and these never change. They bind to the order of the cubes that are presented to `Resolve.__init__` or `Resolve.__call__`. In the context of cube arithmetic, these are the `lhs` and the `rhs` operands of the binary arithmetic operation. It's important to maintain the order of the cubes that are given to `Resolve`, and not swap them around. 

During the process of preparing to reconcile or resolve the `lhs` and `rhs` cubes into a single resultant cube, one cube must map to the other i.e., you could think of it as, one of the cubes will be the target for an in-place operation. Typically, it's the cube with the higher dimensionality, in that case that's the `tgt` and the other is the `src`, but this may change due to broadcasting and free dimensions (i.e., anonymous dimensions with no coordinate metadata describing them).

Also, this situation gets more involved when the `lhs` and `rhs` have the same dimensionality... the point is, the `src` and `tgt` are not akin to the `lhs` and `rhs`, as the mapping of `src` to `tgt` can change depending on several factors which are determined at runtime as the `lhs` and `rhs` are compared and analysed.

Reply:
@lbdreyer Also, the `Resolve.mapping`, is tightly coupled to the concept of `src` and `tgt`, where the `Resolve.mapping.keys()` are the `src` cube dimensions, and the `Resolve.mapping.values()` are the `tgt` cube dimensions, always.
The freedom and flexibility to invert this `Resolve.mapping` relationship, which is required at appropriate times, is totally independent of the `lhs` and `rhs` original operation ordering.
That's a simple, yet important concept, that makes quite complex behaviour simply possible.